desc/vmec_utils.py

```diff
@@ -469,3 +487,509 @@ def vmec_boundary_subspace(eq, RBC=None, ZBS=None, RBS=None, ZBC=None):  # noqa:
+    M_booz : int, optional
+        poloidal resolution to use for Boozer transform.
+    N_booz : int, optional
+        toroidal resolution to use for Boozer transform.
+    verbose: int
+        Level of output (Default = 1).
+        * 0: no output
+        * 1: status of quantities computed
+        * 2: as above plus timing information
+
+    Returns
+    -------
+    None
+
+    """
+    timer = Timer()
+    timer.start("Total time")
+
+    Psi = eq.Psi
+    NFP = eq.NFP
+    if M_booz is None:
+        M_booz = 2 * eq.M
+    if N_booz is None:
+        N_booz = 2 * eq.N
+
+    # calculations are done on a "half-grid" of size surfs-1
+    # which starts at s =  psi = 0.5/(surfs-1)
+    # and increments by 1 / (surfs-1)
+    # until it reaches 1-0.5/(surfs-1)
+    # VMEC radial coordinate: s = rho^2 = Psi / Psi(LCFS)
+    s_full = np.linspace(0, 1, surfs)
+    hs = 1 / (surfs - 1)
+    s_half = np.arange(hs / 2, 1, hs)
+    r_full = np.sqrt(s_full)
+    r_half = np.sqrt(s_half)
+
+    basis = get_transforms(
+        "|B|_mn",
+        obj=eq,
+        grid=Grid(np.array([])),
+        M_booz=M_booz - 1,
+        N_booz=N_booz,
+    )["B"].basis
+
+    matrix, modes = ptolemy_linear_transform(basis.modes)
+    num_modes = modes.shape[0]
+
+    if eq.sym:  # need a separate sin basis for Z (maybe nu)
+        basis_sin = get_transforms(
+            "|B|_mn",
+            obj=eq,
+            grid=Grid(np.array([])),
+            M_booz=M_booz - 1,
+            N_booz=N_booz,
+            sym="sin",
+        )["B"].basis
+        matrix_sin, modes_sin = ptolemy_linear_transform(
+            np.insert(basis_sin.modes, 0, np.array([0, 0, 0]), axis=0)
+        )
+        modes_sin[0, 0] = -1
+        # make the first mode a "sin" mode, even though it is m=n=0
+        # because in the VMEC style outputs it includes this mode
+
+    else:
+        matrix_sin = matrix
+        modes_sin = modes
+
+    timer.start("Boozer Transform")
+
+    B_mn = np.array([[]])
+    R_mn = np.array([[]])
+    Z_mn = np.array([[]])
+    Nu_mn = np.array([[]])
+    Sqrt_g_B_mn = np.array([[]])
+
+    vol_grid = LinearGrid(M=2 * M_booz, N=2 * N_booz, NFP=eq.NFP, rho=r_half)
+    # precompute the needed data for the boozer surface computations
+    # (except those which have to be computed on a single surface only,
+    # like theta_B, zeta_B or nu)
+    keys = ["|B|", "R", "Z", "sqrt(g)", "rho", "psi_r", "lambda", "B_zeta", "B_theta"]
+    data_vol = eq.compute(
+        keys,
+        grid=vol_grid,
+    )
+
+    # FourierZernike fit the B_theta and B_zeta, then at each surface
+    # just evaluate the fit at the rho to get the B_theta_mn and B_zeta_mn
+    # so this we only only do a single fit
+    transform = Transform(
+        grid=vol_grid,
+        basis=FourierZernikeBasis(
+            L=2 * eq.L, M=M_booz - 1, N=N_booz, sym="cos", NFP=eq.NFP
+        ),
+        build_pinv=True,
+    )
+    B_zeta_lmn = transform.fit(data_vol["B_zeta"])
```

Comment:
are we sure B_zeta and B_theta are analytic quantities? B obviously is, but i thought the basis vectors aren't at the axis which would make the components possibly not analytic

Reply:
I am not certain, but this seems to work out fine and agree basically exactly with just calculating B_zeta_mn on a surface

Reply:
You were right, I must have made a mistake when I tested that initially, it is inaccurate to do it the 3D way. I reverted it back to calculating it surface by surface